File: app/src/main/java/com/example/im/service/ImCenter.java
package com.example.im.service;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;

public class ImCenter {
    private WebSocketClient mClient;
    private OnLogListener mLogListener;

    public ImCenter(String ip) {
        URI uri = URI.create(ip);
        mClient = new WebSocketClient(uri) {
            @Override
            public void onOpen(ServerHandshake serverHandshake) {
                log(serverHandshake.toString());
            }

            @Override
            public void onMessage(String s) {
                log(s);
            }

            @Override
            public void onClose(int i, String s, boolean b) {
                log(s + "|" + i + "|" + b);
            }

            @Override
            public void onError(Exception e) {
                log(e.getMessage());
            }
        };
    }

    public void setLogListener(OnLogListener listener) {
        mLogListener = listener;
    }

    public void connect() {
        if (mClient.isClosed()) {
            mClient.connect();
        }
    }

    public void disconnect() {
        if (mClient.isOpen()) {
            mClient.close();
        }
    }

    public void send(MessageSending msg) {
        mClient.send(msg.toString());
    }

    private void log(String content) {
        if (mLogListener != null) {
            mLogListener.log(content);
        }
    }
}

File: app/src/main/java/com/example/im/service/ImCenter.kt
package com.example.im.service

import com.example.im.MainHandler
import com.example.im.model.MessageSending
import com.example.im.model.OnLogListener
import org.java_websocket.client.WebSocketClient
import org.java_websocket.handshake.ServerHandshake
import java.net.URI

class ImCenter(ip: String) {
    private var mClient: WebSocketClient? = null
    private var mLogListener: OnLogListener? = null

    init {
        val uri = URI.create(ip)
        try {
            mClient = object : WebSocketClient(uri) {
                override fun onOpen(serverHandshake: ServerHandshake) {
                    log(serverHandshake.toString())
                }

                override fun onMessage(s: String) {
                    log(s)
                }

                override fun onClose(i: Int, s: String, b: Boolean) {
                    log("$s|$i|$b")
                }

                override fun onError(e: Exception) {
                    log(e.message)
                }
            }
        } catch (ex: Exception) {
            log("Exception:" + ex.message)
        }

    }

    fun setLogListener(listener: OnLogListener) {
        mLogListener = listener
    }

    fun connect() {
        if (!mClient!!.isOpen) {
            mClient!!.connect()
        }
    }

    fun disconnect() {
        if (mClient!!.isOpen) {
            mClient!!.close()
        }
    }

    fun send(msg: MessageSending) {
        if (mClient!!.isOpen) {
            mClient!!.send(msg.toString())
        }
    }

    private fun log(content: String?) {
        if (content != null && mLogListener != null) {
            MainHandler.main(Runnable { mLogListener!!.log(content) })
        }
    }
}
